### ISSUE TITLE
Introducing OPENTOFU_ environment variables

### ISSUE THREAD

**Issue by dvaumoron** [opentofu/opentofu#641]
### Summary

The goal is to duplicate all `TF_` environment variables to `OPENTOFU_`, with priority to `OPENTOFU_` version and possibly a warning to encourage the migration when `TF_` one are set.

### Problem Statement

Using `OPENTOFU_` prefixed environment variables for new functionnalities is a good way to distinguish from the predecessor.

However duplication instead of renaming allows to stay in the spirit of [compatibility promises](https://opentofu.org/docs/language/v1-compatibility-promises/) for existing environment variables.

### User-facing description

with `TF_TEMP_LOG_PATH` set and no `OPENTOFU_TEMP_LOG_PATH` display a message to encorage the use of `OPENTOFU_TEMP_LOG_PATH` and work with the value from `TF_TEMP_LOG_PATH`

with `OPENTOFU_TEMP_LOG_PATH` set, no particular display and work with the value from `OPENTOFU_TEMP_LOG_PATH` (whether the value of `TF_TEMP_LOG_PATH` is defined or not)

### Technical Description

I think a simple way to do this is to replace all the call `os.Getenv("TF_*")` with a custom `Getenv` able to handle the the two version of the environment variable (simple string manipulation which will depend on the chosen format of the input string (prefixed or not)), all the constant `TF_*` will be changed too.

### Rationale and alternatives

If we don't do this feature, user will probably be confused between old `TF_` environment variable and new one using the `OPENTOFU_` prefix.

An alternative is to stick to `TF_` everywhere but it will not allow a clear distinction on new features included in OpenTofu
 

### Related Issues

This RFC follows discussions in #618

**Comment by dvaumoron** [comment]
I forgot to mention in "Technical Description" : added environment variable will need to use the custom `Getenv` only when the addition is meant to stay compatible with an evolution in TF

**Comment by davidjeddy** [comment]
Yes and please. Really appreciate the consideration to not simply find/replace. Keeping the TF_* vars and outputting a WARN or INFO about migrating to OPENTOFU_* is wonderful. Organizations can migrate to OpenTofu w/o breaking the manymanymanymany automations currently in place.

**Comment by charmitro** [comment]
This is indeed a very useful feature. To quote @davidjeddy, it will help organizations considering a migration to OpenTofu worry less about potential catastrophic events.

I'm also willing to implement it if accepted - as my first contribution.

**Comment by Hefeweizen** [comment]
> Afterthought : the same custom version of `Getenv` can contains the logic detecting a `TF_` prefix to check the `OPENTOFU_` version in that case, and else just call `os.Getenv`, this way there is no need to change the existing constant and the custom version can be used everywhere (with only a small overhead)

Just calling out:  +1 to the abstracted handling of the env_vars, but we should be careful that we only allow multiple versions of the legacy env_vars, not the future features we're adding.  That is, if we're creating `OPENTOFU_NEW_FEATURE_1` we don't want to inadvertently honor `TF_NEW_FEATURE_1`.  Doing so erodes some of separation we've achieved by introducing a new variable namespace.

**Comment by dvaumoron** [comment]
I hope making it clear in the function comment will be enough to avoid most mistakes (it appears in IDE and godoc)

**Comment by StephanHCB** [comment]
Please consider logging INFO instead of WARN if a legacy TF_* environment variable is used.

IMHO, if opentofu ever were to decide to deprecate the legacy variables, _then_ it should become a warning.

**Comment by skyzyx** [comment]
I feel that this aligns well with the work to add support for `.tofu` files.

Sometimes, you need to configure OpenTofu slightly differently from Terraform — especially when supporting both — and this helps address that use-case.

**Comment by ucfchandra** [comment]
Just want to make a quick comment that I would prefer TOFU_ instead of OPENTOFU_ for ergonomic reasons

**Comment by skyzyx** [comment]
@hoppeeri said:

> > Just want to make a quick comment that I would prefer TOFU_ instead of OPENTOFU_ for ergonomic reasons
> 
> why not OT_ instead? Doesn't even break string length compatibility.

IMO, `OT` is just a bit _too_ ambiguous. At-a-glance, it's tougher to interpret as being OpenTofu-related than `TOFU` or `OPENTOFU` — one of the _benefits_ of having such an odd name.

The two hardest problems in computer science are:

1. Cache invalidation.
2. Naming things.
3. Off-by-one errors.

As a former UX practitioner, I stand strongly by "clarity over cleverness (or succinctness)."
